### DISCUSSION TITLE
How to write a file during build

### DISCUSSION THREAD

**Discussion by Svish** [vercel/next.js#60462] [Help]
### Summary

I want to write a file with some content during build, so that I can read it back later and use it for various purposes in various routes, both static and dynamic ones. It needs to work when running `next dev`, but even more importantly it also needs to work when built and deployed by Vercel.

Some usage examples:

* Create an index of all the frontmatter of many mdx-files, so I can create lists, sitemaps, feeds, and so on, without having to read and parse _all_ the mdx-files every single time.
* Store the output of a custom command or node.js script in a file, for example to generate some random fake data, store the output of some program, etc.

So...

1. What's the best/correct place to "hook in" to the nextjs/vercel build process for this?
2. What's the best/correct place to write the file, so that it's available when the website is actually up and running?

I'm not talking about writing files while the site is live btw., which I'm not even sure would/should work at all with how deployments on vercel (and other places) work with edge functions and all that? I just want to write write some files during the _build_, based on other files in the project or command output, that I want to be available to server-side code when the website is deployed.

### Additional information

```js
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro   
Binaries:
  Node: 20.4.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.0.4
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

**Comment by icyJoseph** [comment]
You could maybe do a `pre` script? for both dev and build? 

- https://docs.npmjs.com/cli/v9/using-npm/scripts#pre--post-scripts

**Reply by Svish** [reply]
If that works on Vercel, then that could be a place to "hook in", yes.

Not sure where I should write the files to though, so that they become a proper part of the deployment.

**Reply by icyJoseph** [reply]
It should work on the build step, yes. 

And probably write them to the public folder. I'll take a look later.

**Reply by Svish** [reply]
Writing it to the public folder would make the file available to everyone though, not just my server-side code, which feels a bit... sketchy?

Is there anywhere else I can write it to? Or if not, can I configure vercel or next.js somehow to "hey, I know this folder is not the public one, but please bring along this _whole_ directory as well"?

**Reply by icyJoseph** [reply]
Well, if you just need it to be available during the build step, you should just be able to generate them through the pre script step. Anywhere within the git project.

**Reply by Svish** [reply]
No, I need it to be available _after_ the build as well.

Let's say we have a bunch of MDX files, which contains a `publishedDate` field as part of the front-matter. Most dates are in the past, but some are in the future and should only show up on the site later when the publish date passes.

I want to create an index of all these MDX files during build, so that it's easy to for example generate a dynamic list of the 10 latest blog posts on a page. It can't be statically generated during build time, since what posts are the 10 latest will change depending on the current date. I also don't want to re-parse all the files to find the latest 10 posts every time I need that list.

**Reply by icyJoseph** [reply]
But, maybe I am missing something here, but you'd output a file, say to `src/pre-calculated/data.json` then, you can do something like `fs.readFile(`${process.cwd()}/pre-calculated/data.json`)` (pseudocode from https://nextjs.org/docs/pages/api-reference/functions/get-static-props#reading-files-use-processcwd) ~ 

The one gotcha is, after the Vercel build step, will the generated file, that's not part of `.git`, be available for the runtime?

**Reply by Svish** [reply]
Yeah, that's what I'm wondering, and was curious if someone from Vercel would have a definitive answer to that, like what is the proper way to do that.

Also kind of hoping that some on the Next team could write some actual documentation for this. I don't understand how I can be the only person who would need something like this.

There's a lot of tutorials on "how to make blog posts with Next", but they all just seem to focus on rendering the blog post itself, not how to actually loop through and gather metadata in a properly efficient way. Or they use tools like `contentlayer` (which I was as well, but it seems to have died now, so I'm trying to figure out how to do things myself in a simple way).

**Comment by don-esteban** [comment]
Maybe I also do not understand well, what you want to achieve. If your MDX index file gets really huge, GBs I mean, you will need an  external datastore. I can't imagine that you have that amount of MDX files. Otherweise import the index to pages from a helper module (business logic module, data layer module, ...).

Something like that in pseudo code:

```
// App router page component

import mdxIndex from '@/lib/data-layer/mdx-index'

export default function SomePage() {
  return mdxIndex.map( (item) => <div key=...>{item.date}</div>
}
```

```
// @/lib/data-layer/mdx-index.js

import fs from 'fs';
const mdxIndex = buildTheIndexSomehow()
export mdxIndex

// Or export a function

export function getPostsBeforeDate(date) {
  return mdxIndex.filter(...)
}
```

This is no code. It's just to illustrate my idea, which came first to my mind.

**Reply by don-esteban** [reply]
Thinking twice, probably this is not what you need. I did a quick check in an Next.js project but it seems that modules get executed multiple times during build. Node generally caches modules, which misleads my initial thoughts.

I still believe for both of your use cases (MDX, OS process etc.) exist better approaches than writing (caching) to the file system. I would go for thinking and reading about the new Next.js caching features.

See: https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating

**Reply by Svish** [reply]
Yeah, actually building the index is not really the issue here, and it will not be enormous, quite small probably. I just want to avoid having to read and parse 200 mdx files each time I want the meta-data.

There might be better approaches, but my issue here is that the Next docs and various tutorials and examples I've seen doesn't mention this at all it seems.

As a temporary solution (but maybe it's the final one?) I actually have tried to use the `unstable_cache` from Next. And it does improve performance quite a lot, but sometimes also not... so I don't quite understand what's going on with that. That's why I'm thinking just writing my own index somehow during build would be much more predictable and understandable in the long run.

**Reply by don-esteban** [reply]
Something I don't understand. You want to generate the index cache during build and use it after build? Or you want to prebuild the index to speed up build time? If your concern is build time, why not just add a prebuild step in your publishing pipeline before pushing the repo to Git?

Perhaps Server Components help? Generate your MDX post index lists and use caching. https://nextjs.org/docs/app/building-your-application/rendering/server-components.

I always try to resort to the basics/standards and don't overcomplicate things. But you have details, you know better. Good luck!

**Reply by Svish** [reply]
Both, speed up build _and_ use it after build in dynamic routes.

Already using RSC, which are __great_, but doesn't help in this regard.

**Comment by don-esteban** [comment]
https://react.dev/reference/react/cache will be helpful too. No need to write own cache files. See https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns too.

**Comment by haideralipunjabi** [comment]
Any way to write files *during* the build. For e.g, list of all images which use the `Image` component using a custom loader. The write call would be part of the custom loader.
